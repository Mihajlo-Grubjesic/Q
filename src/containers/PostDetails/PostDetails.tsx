import { useParams } from "react-router-dom";
import { Button } from "../../components/Button/Button";
import { ContentWrapper } from "../../components/Card/styled/Card";
import { CardDetails } from "../../components/CardDetails/CardDetails";
import { NavigationLink } from "../../components/Link/Link";
import { ROUTES } from "../../constants/routes";
import { PostData } from "../../types";

interface Props {
  posts: PostData[];
}

export const PostDetails = ({ posts }: Props): JSX.Element => {
  const { postId } = useParams();
  const post = posts.find((post) => post.id.toString() === postId);
  return (
    <>
      <CardDetails
        title={post?.title}
        subtitle={post?.username}
        content={
          <ContentWrapper>
            {post?.comments.map((comment) => (
              <p key={comment}>* {comment}</p>
            ))}
          </ContentWrapper>
        }
      />
      <NavigationLink to={ROUTES.POSTS}>
        <Button>‚Üê Go Back</Button>
      </NavigationLink>
    </>
  );
};
