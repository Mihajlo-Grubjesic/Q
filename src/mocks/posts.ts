import { Post } from '../types';

export const PostsMock: Post[] = [
  {
    id: 1,
    userId: 1,
    body: 'Post 1',
    title: 'Post 1',
  },
  {
    id: 2,
    userId: 2,
    body: 'Post 2',
    title: 'Post 2',
  },
  {
    id: 3,
    userId: 3,
    body: 'Post 3',
    title: 'Post 3',
  },
  {
    id: 4,
    userId: 4,
    body: 'Post 4',
    title: 'Post 4',
  },
];
