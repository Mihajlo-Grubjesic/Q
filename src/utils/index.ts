import { Post, User, Comment, PostData } from "../types";

export const combinePostsData = ({
  posts,
  users,
  comments,
}: {
  posts: Post[];
  users: User[];
  comments: Comment[];
}): PostData[] => {
  return posts.map((post) => {
    const username =
      users.find((user) => user.id === post.userId)?.username ?? "";
    const postComments = comments
      .filter((comment) => comment.postId === post.id)
      .map((comment) => comment.body);
    return { ...post, username, comments: postComments };
  });
};
