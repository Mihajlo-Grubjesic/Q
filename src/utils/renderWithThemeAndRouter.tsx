import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { DefaultTheme, ThemeProvider } from 'styled-components';
import { lightTheme } from '../theme/themes';

export function renderWithThemeAndRouter(
  children: React.ReactElement,
  theme?: DefaultTheme,
) {
  return render(
    <ThemeProvider theme={theme ? theme : lightTheme}>
      <Router>{children}</Router>
    </ThemeProvider>,
  );
}
