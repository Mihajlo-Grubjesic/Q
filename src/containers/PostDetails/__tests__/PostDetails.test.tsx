import { vi } from 'vitest';
import { fireEvent, screen } from '@testing-library/react';
import { renderWithThemeAndRouter } from '../../../utils/renderWithThemeAndRouter';

import { CommentsMock } from '../../../mocks/comments';
import { PostsMock } from '../../../mocks/posts';
import { UsersMock } from '../../../mocks/users';
import { combinePostsData } from '../../../utils';
import PostDetails from '../PostDetails';

const mockUseParams = vi.fn().mockReturnValue({
  postId: '1',
});

vi.mock('react-router-dom', async () => {
  const actual: object = await vi.importActual('react-router-dom');
  return { ...actual, useParams: () => mockUseParams() };
});

vi.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

describe('PostDetails', () => {
  it('renders post details', () => {
    const posts = combinePostsData({
      posts: PostsMock,
      users: UsersMock,
      comments: CommentsMock,
    });
    const { container } = renderWithThemeAndRouter(
      <PostDetails posts={posts} />,
    );
    expect(container).toHaveTextContent('Post 1');
    const backButton = screen.getByRole('button', {
      name: /goBack/i,
    });
    expect(backButton).toBeInTheDocument();
    fireEvent.click(backButton);
    expect(window.location.pathname).toEqual('/posts');
    expect(container).toMatchSnapshot();
  });
});
