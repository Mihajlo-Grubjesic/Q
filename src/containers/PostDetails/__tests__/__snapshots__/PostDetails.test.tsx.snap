// Vitest Snapshot v1

exports[`PostDetails > renders post details 1`] = `
<div>
  <div
    class="sc-gKPRtg iapbnP"
  >
    <div
      class="sc-eDvSVe fXzfkg"
    >
      Post 1
    </div>
    <div
      class="sc-jSUZER dDsNen"
    >
      Bret
    </div>
    <div
      class="sc-dkrFOg ftGKqN"
    >
      <p>
        ● 
        laudantium enim quasi est quidem magnam voluptate ipsam eos
tempora quo necessitatibus
dolor quam autem quasi
reiciendis et nam sapiente accusantium
      </p>
      <p>
        ● 
        est natus enim nihil est dolore omnis voluptatem numquam
et omnis occaecati quod ullam at
voluptatem error expedita pariatur
nihil sint nostrum voluptatem reiciendis et
      </p>
      <p>
        ● 
        quia molestiae reprehenderit quasi aspernatur
aut expedita occaecati aliquam eveniet laudantium
omnis quibusdam delectus saepe quia accusamus maiores nam est
cum et ducimus et vero voluptates excepturi deleniti ratione
      </p>
      <p>
        ● 
        non et atque
occaecati deserunt quas accusantium unde odit nobis qui voluptatem
quia voluptas consequuntur itaque dolor
et qui rerum deleniti ut occaecati
      </p>
      <p>
        ● 
        harum non quasi et ratione
tempore iure ex voluptates in ratione
harum architecto fugit inventore cupiditate
voluptates magni quo et
      </p>
    </div>
  </div>
  <a
    class="sc-iBYQkv kGBVhA"
    href="/posts"
  >
    <button
      class="sc-bcXHqe fetPcC"
    >
      ← goBack
    </button>
  </a>
</div>
`;
